// @sanity/presentation/dist/_chunks-cjs/PostMessagePreviewSnapshots.cjs
"use strict";
var react = require("react"), rxjs = require("rxjs"), sanity = require("sanity");
const PostMessagePreviews = (props) => {
  const {
    comlink,
    refs,
    perspective
  } = props, documentPreviewStore = sanity.useDocumentPreviewStore(), schema = sanity.useSchema(), refsSubject = react.useMemo(() => new rxjs.Subject(), []), previews$ = react.useMemo(() => refsSubject.asObservable().pipe(rxjs.switchMap((refs_0) => rxjs.combineLatest(refs_0.map((ref) => {
    const draftRef = {
      ...ref,
      _id: sanity.getDraftId(ref._id)
    }, draft$ = perspective === "previewDrafts" ? documentPreviewStore.observeForPreview(draftRef, schema.get(draftRef._type)).pipe(
      // Share to prevent double subscribe in the merge
      rxjs.share(),
      // Don't emit if no snapshot is returned
      rxjs.skipWhile((p) => p.snapshot === null)
    ) : (
      // Don't emit if not displaying drafts
      rxjs.NEVER
    ), publishedRef = {
      ...ref,
      _id: sanity.getPublishedId(ref._id)
    }, published$ = documentPreviewStore.observeForPreview(publishedRef, schema.get(publishedRef._type));
    return rxjs.merge(published$.pipe(rxjs.takeUntil(draft$)), draft$).pipe(rxjs.filter((p_0) => !!p_0.snapshot), rxjs.map((p_1) => {
      const snapshot = p_1.snapshot;
      return {
        _id: sanity.getPublishedId(snapshot._id),
        title: snapshot.title,
        subtitle: snapshot.subtitle,
        description: snapshot.description,
        imageUrl: snapshot.imageUrl
      };
    }));
  }))), rxjs.debounceTime(0)), [documentPreviewStore, refsSubject, schema, perspective]);
  return react.useEffect(() => {
    const sub = previews$.subscribe((snapshots) => {
      comlink.post({
        type: "presentation/preview-snapshots",
        data: {
          snapshots
        }
      });
    });
    return () => {
      sub.unsubscribe();
    };
  }, [comlink, previews$]), react.useEffect(() => {
    refsSubject.next(refs);
  }, [refs, refsSubject]), null;
};
var PostMessagePreviewSnapshots = react.memo(PostMessagePreviews);
exports.default = PostMessagePreviewSnapshots;
